privacyidea/api/lib/prepolicy.py

```diff
@@ -131,9 +131,44 @@ def policy_wrapper(*args, **kwds):
         return policy_wrapper
 
 
+def set_random_pin(request=None, action=None):
+    """
+    This policy function is to be used as a decorator in the API setrandompin function
+    If the policy is set accordingly it adds a random PIN to the
+    request.all_data like.
+
+    It uses the policy SCOPE.ENROLL, ACTION.OTPPINSETRANDOM to set a random OTP
+    PIN during Token enrollment
+    """
+    params = request.all_data
+    policy_object = g.policy_object
+    # Fixme: we need to allow passing the user
+    #user_object = get_user_from_param(params)
+    # get the length of the random PIN from the policies
+    adminname = g.logged_in_user.get("username")
+    adminrealm = g.logged_in_user.get("realm")
+    pin_pols = policy_object.get_action_values(action=ACTION.OTPPINSETRANDOM,
```

Comment:
Correct me if I'm wrong, but wouldn't this give wrong results if the currently logged-in user is a user, not an admin? In this case, we would match the policy attribute ``adminrealm`` against the user's realm.

Reply:
I just noticed that the docstring says ``OTPPINSETRANDOM`` is in scope=ENROLL instead of scope=ADMIN -- I think this would make more sense?

Reply:
A user can not set a random pin.

Reply:
done in a2785dfe

Reply:
Hm, I'm not sure I understand: Doesn't ``user.rst`` say that 

> If the ``setrandompin`` action is defined, the user
> is allowed to call the endpoint, that sets a random PIN on his
> specified token.

?

Reply:
Yes it says. and lib/policy.py also allows the user to do so.

Reply:
Easy, so we need to distinguish the loggedin role here.

Reply:
done in 83a95f73
(I think I will not write an additional test ;-)

Reply:
Sorry, I'm still not sure this is correct: Now, if a user is logged in and uses the ``/token/setrandompin`` endpoint, the decorator checks for a scope=USER, ``otp_pin_set_random`` policy action. But it is impossible to define such a policy in the web UI. :/

Do we even want a policy action ``otp_pin_set_random`` in scope=ADMIN as well as in scope=USER?

Reply:
otp_pin_set_random defines the length of the OTP PIN, right?

In the scope user we have a policy "setrandompin". So I think we allow the user for whatever reasons to set a random pin on his tokens.

But: Then we are simply missing the policy ``otp_pin_set_random`` in the policy definition. I would simply add it.

Reply:
OK, I guess that makes sense. So this would even allow to configure different lengths for random PINs for users and admins.

Reply:
done in 83a95f73

Reply:
thanks! I think that's nicer for consistency.
Right now, we still have the behavior that without any user policies, the "set random PIN" button is displayed, but does not work without a "otp_pin_set_random" policy. Should we keep it like that, or should we set a default length?